fedimint-metrics/src/main.rs

```diff
@@ -0,0 +1,419 @@
+                .local();
+            for m in ln_funded_contract_outgoing_account_amounts_sats {
+                local.observe(m.msats as f64 / 1000.0)
+            }
+
+            metrics
+                .process_update_time_ms
+                .set(start.elapsed()?.as_millis().try_into()?);
+            state.last_update_time = Some(start);
+        }
+        let mut interval = Duration::from_secs(args.update_interval);
+        interval = interval.saturating_sub(start.elapsed()?);
+        debug!("Sleeping for {interval:?}");
+        fedimint_core::task::sleep(interval).await;
+    }
+}
+
+async fn get_metrics(
+    axum::extract::State(state): axum::extract::State<Arc<RwLock<State>>>,
+    args: MonitorArgs,
+) -> (StatusCode, String) {
+    let state = state.read().await;
+    if state.last_update_time.is_none()
+        || fedimint_core::time::now()
+            .duration_since(state.last_update_time.unwrap())
+            .expect("time to work")
+            > Duration::from_secs(args.update_interval) * 2
+    {
+        return (
+            StatusCode::INTERNAL_SERVER_ERROR,
+            format!(
+                "Metrics are outdated, last updated time was: {:?}",
+                state.last_update_time,
+            ),
+        );
+    }
+    let metric_families = state.metrics.registry.gather();
+    let result = || -> anyhow::Result<String> {
+        let mut buffer = Vec::new();
+        let encoder = TextEncoder::new();
+        encoder.encode(&metric_families, &mut buffer)?;
+        Ok(String::from_utf8(buffer)?)
+    };
+    match result() {
+        Ok(result) => (StatusCode::OK, result),
+        Err(e) => (StatusCode::INTERNAL_SERVER_ERROR, format!("{e:?}")),
+    }
+}
+
+#[derive(Default)]
+struct State {
+    metrics: PrometheusMetrics,
+    last_update_time: Option<SystemTime>,
+}
+
+type U64IntGauge = prometheus::core::GenericGauge<prometheus::core::AtomicU64>;
+
+#[derive(Debug, Clone)]
+struct PrometheusMetrics {
+    registry: Registry,
+    ln_incoming_offer: U64IntGauge,
+    ln_contract_outcome_incoming: U64IntGauge,
+    ln_contract_outcome_outgoing: U64IntGauge,
+    ln_output_outcome_offer: U64IntGauge,
+    ln_output_outcome_cancel_outgoing_contract: U64IntGauge,
+    ln_funded_contract_incoming: U64IntGauge,
+    ln_funded_contract_outgoing: U64IntGauge,
+    ln_funded_contract_outgoing_cancelled: U64IntGauge,
+    ln_funded_contract_incoming_account_amounts_sats: Histogram,
+    ln_funded_contract_outgoing_account_amounts_sats: Histogram,
+    ln_funded_contract_outgoing_cancelled_account_amounts_sats: Histogram,
+    process_update_time_ms: U64IntGauge,
+}
+
+impl PrometheusMetrics {
+    pub fn new() -> anyhow::Result<Self> {
+        let registry = Registry::new();
+
+        let ln_incoming_offer =
+            U64IntGauge::new("ln_incoming_offer", "contracts::IncomingContractOffer")?;
+        registry.register(Box::new(ln_incoming_offer.clone()))?;
```

Comment:
Why does this require `Box::new`?

Reply:
Sometimes you have many implementations of a trait and using generics can be tricky, then you alloc on the heap with box and things are easier to implement/use.